### PR TITLE
Add support for NoNewPrivileges

### PR DESCRIPTION
**Pull request by mrunalp**
Support NoNewPrivileges for additional security in containers.

### CODE REVIEW

**Comment by mrunalp** [comment]
@LK4D4 @crosbymichael ping
cc: @rhatdan

File: spec.go [hunk 1]
```diff
@@ -300,6 +301,7 @@ func createLibcontainerConfig(cgroupName string, spec *specs.LinuxSpec) (*config
 	config.Sysctl = spec.Linux.Sysctl
 	config.ProcessLabel = spec.Linux.SelinuxProcessLabel
 	config.AppArmorProfile = spec.Linux.ApparmorProfile
+	config.NoNewPrivileges = spec.Linux.NoNewPrivileges
```
Comment (cyphar) [inline spec.go]: Is there a spec update PR for this?

Comment (mrunalp) [inline spec.go]: This was pulled in a previous spec update.

**Comment by rhatdan** [comment]
This looks got to me.  Need to get this into Docker ASAP, since it really helps people running non privileged work loads.  Eliminates a lot of worry about contained non priv processes getting access to setuid or setfcap files.

File: libcontainer/configs/config.go [hunk 2]
```diff
@@ -171,6 +171,9 @@ type Config struct {
 	// A default action to be taken if no rules match is also given.
 	Seccomp *Seccomp `json:"seccomp"`
 
+	// NoNewPrivileges controls whether processes in the container can gain additional privileges.
+	NoNewPrivileges bool `json:"nonewprivileges"`
```
Comment (hqhq) [inline libcontainer/configs/config.go]: Not a problem in this PR.
For json property names, we have `nonewprivileges` and `noNewPrivileges` and `no_new_privileges` in code, looks like we don't have code style for that yet, should we change them for consistency.

File: libcontainer/setns_init_linux.go [hunk 3]
```diff
@@ -45,5 +45,10 @@ func (l *linuxSetnsInit) Init() error {
 			return err
 		}
 	}
+	if l.config.Config.NoNewPrivileges {
+		if err := system.Prctl(PR_SET_NO_NEW_PRIVS, 1, 0, 0, 0); err != nil {
+			return err
+		}
+	}
```
Comment (hqhq) [inline libcontainer/setns_init_linux.go]: I think we should do this before seccomp rules have been applied.

Comment (cyphar) [inline libcontainer/setns_init_linux.go]: And, if possible, do it before we sync with the parent.

File: libcontainer/configs/config.go (hunk 2 above)
Comment (mrunalp) [inline libcontainer/configs/config.go]: Yeah, sticking to whatever style we have in each repo for now :/

File: libcontainer/setns_init_linux.go (hunk 3 above)
Comment (mrunalp) [inline libcontainer/setns_init_linux.go]: Moved before applying seccomp settings.

**Comment by mrunalp** [comment]
Updated.

**Comment by rhatdan** [comment]
Are we sure this work? Can you still apply seccomp rules after you have set NO_NEW_PRIVS?  I thought I heard someone say that you could not.

File: libcontainer/standard_init_linux.go [hunk 4]
```diff
@@ -109,6 +113,11 @@ func (l *linuxStandardInit) Init() error {
 	if err := syncParentReady(l.pipe); err != nil {
 		return err
 	}
+	if l.config.Config.NoNewPrivileges {
```
Comment (cyphar) [inline libcontainer/standard_init_linux.go]: If there's no reason to place it after `syncParentReady` (like `syncParentReady` might break it), I'd prefer to place it before. Go likes to create new processes each time you do a syscall, which is the whole point of late cgroup apply (applying cgroups as late as possible means we're less likely to hit our resource limits during setup).

**Comment by cyphar** [comment]
@rhatdan I don't think that's right, if I'm reading [this LWN article][lwn] correctly:

> Lutomirski also notes that a sandbox will be much less useful if execve() has to fail when there is any kind of security transition, as Cox suggested. The presence of a policy on a particular binary would make that binary unusable from within a sandbox, no matter what the policy is. A better solution, Lutomirski said, is to set the no_new_privs bit, then set up a sandbox (using Drewry's seccomp system call filtering for example), then execute the binary:

> > If you want a sandbox, call PR_SET_NO_NEW_PRIVS, then enable seccomp (or whatever) to disable ptrace, evil file access, connections on unix sockets that authenticate via uid, etc. 

[lwn]: https://lwn.net/Articles/475678/

**Comment by rhatdan** [comment]
Great.  I am probably misremembering.

**Comment by mrunalp** [comment]
Yes, I had looked at libseccomp earlier and it makes the call as well
https://github.com/seccomp/libseccomp/blob/7f3ae6e6a12390bd38f0787b242f60c47ad076c3/src/system.c#L75

**Comment by mrunalp** [comment]
I checked that there is no harm in this getting called twice but prctl itself could get blocked by seccomp so makes sense to move this before seccomp.

File: libcontainer/standard_init_linux.go (hunk 4 above)
Comment (mrunalp) [inline libcontainer/standard_init_linux.go]: Moved it above the sync.

**Comment by crosbymichael** [comment]
LGTM

**Comment by LK4D4** [comment]
LGTM

**Pull request merged**